edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,244 @@
+            {
+                // Update case - this is just remove + re-insert
+                this.Remove(identity.Id);
+            }
+
+            // Insert case
+            if (identity.IsModule)
+            {
+                this.InsertModuleIdentity(writableSnapshot, identity);
+            }
+            else
+            {
+                this.InsertDeviceIdentity(writableSnapshot, identity);
+            }
+
+            // Write back to the atomic field
+            if (!this.nodes.CompareAndSet(snapshot, writableSnapshot.ToImmutableDictionary()))
+            {
+                // There's only ever one thread calling InsertOrUpdate, so this should never fail
+                throw new InvalidOperationException("Multi-threaded insert/update detected on ServiceIdentityTree");
+            }
+        }
+
+        public void Remove(string id)
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            var writableSnapshot = new Dictionary<string, ServiceIdentityTreeNode>(snapshot);
+
+            if (writableSnapshot.TryGetValue(id, out ServiceIdentityTreeNode target))
+            {
+                // Unhook from the parent
+                target.Parent.ForEach(parent => parent.Children.Remove(target));
+
+                // Unhook the children
+                target.Children.ForEach(child => child.Parent = Option.None<ServiceIdentityTreeNode>());
+
+                // Remove
+                writableSnapshot.Remove(id);
+            }
+
+            // Write back to the atomic field
+            if (!this.nodes.CompareAndSet(snapshot, writableSnapshot.ToImmutableDictionary()))
+            {
+                // There's only ever one thread calling Remove(), so this should never fail
+                throw new InvalidOperationException("Multi-threaded delete detected on ServiceIdentityTree");
+            }
+        }
+
+        public bool Contains(string id)
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.ContainsKey(id);
+        }
+
+        public Option<ServiceIdentity> Get(string id)
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.TryGetValue(id, out ServiceIdentityTreeNode treeNode)
+                    ? Option.Some(treeNode.Identity)
+                    : Option.None<ServiceIdentity>();
+        }
+
+        public IList<string> GetAllIds()
+        {
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            return snapshot.Select(kvp => kvp.Value.Identity.Id).ToList();
+        }
+
+        public bool TryGetAuthChain(string id, out string authChain)
+        {
+            Preconditions.CheckNonWhiteSpace(id, nameof(id));
+            authChain = string.Empty;
+
+            // Auth-chain for root
+            if (this.rootDeviceId == id)
```

Comment:
Nit: It may be possible to remove this special case by making this the while loop condition.